src/streams.h

```diff
@@ -13,36 +13,53 @@
 
 #include <algorithm>
 #include <assert.h>
+#include <crypto/common.h>
 #include <cstddef>
 #include <cstdio>
+#include <cstring>
 #include <ios>
 #include <limits>
 #include <optional>
+#include <span>
 #include <stdint.h>
 #include <string.h>
 #include <string>
-#include <utility>
 #include <vector>
 
 namespace util {
-inline void Xor(Span<std::byte> write, Span<const std::byte> key, size_t key_offset = 0)
+inline void XorInt(Span<std::byte> write, const uint64_t key, const size_t size)
 {
-    if (key.size() == 0) {
-        return;
+    assert(size <= write.size());
+    uint64_t raw = 0;
+    memcpy(&raw, write.data(), size);
+    raw ^= key;
+    memcpy(write.data(), &raw, size);
+}
+inline void Xor(Span<std::byte> write, const uint64_t key)
+{
+    assert(key);
+    for (; write.size() >= 8; write = write.subspan(8)) {
+        XorInt(write, key, 8);
     }
-    key_offset %= key.size();
-
-    for (size_t i = 0, j = key_offset; i != write.size(); i++) {
-        write[i] ^= key[j++];
-
-        // This potentially acts on very many bytes of data, so it's
-        // important that we calculate `j`, i.e. the `key` index in this
-        // way instead of doing a %, which would effectively be a division
-        // for each byte Xor'd -- much slower than need be.
-        if (j == key.size())
-            j = 0;
+    switch (write.size()) { // Specify the optimization categories
+    case 0: break;
+    case 1: XorInt(write, key, 1); break;
+    case 2: XorInt(write, key, 2); break;
+    case 4: XorInt(write, key, 4); break;
+    default: XorInt(write, key, write.size());
```

Comment:
Care to elaborate why we don't just have the `default` statement?

Reply:
"Specify the optimization categories" means that the compiler will be able to optimize the cases where one of the parameters (the size) is a constant separately from each other. The default statement would work, but would be very slow, since the 1, 2 and 4 byte versions won't be specialized.

Reply:
How about
```suggestion
    // Help optimizers along by sending constant parameter values into the inlined function,
    // resulting in more efficient substitutions of memcpy() -> native pow-2 copy instructions.
    switch (write.size()) {
    case 0: break;
    case 1: XorInt(write, key, 1); break;
    case 2: XorInt(write, key, 2); break;
    case 4: XorInt(write, key, 4); break;
    default: XorInt(write, key, write.size());
```

Reply:
I ended up with only `// Help the compiler specialize 1, 2 and 4 byte cases`, since the rest was just speculation